sql/query/expr.go

```diff
@@ -0,0 +1,651 @@
+// parse tree. The only significant complexity is the handling of types and
+// implicit conversions. See prepareComparisonArgs and prepareBinaryArgs for
+// details.
+func EvalExpr(expr parser.Expr, env Env) (sqlwire.Datum, error) {
+	switch t := expr.(type) {
+	case *parser.AndExpr:
+		return evalAndExpr(t, env)
+
+	case *parser.OrExpr:
+		return evalOrExpr(t, env)
+
+	case *parser.NotExpr:
+		return evalNotExpr(t, env)
+
+	case *parser.ParenBoolExpr:
+		return EvalExpr(t.Expr, env)
+
+	case *parser.ComparisonExpr:
+		return evalComparisonExpr(t, env)
+
+	case *parser.RangeCond:
+		return evalRangeCond(t, env)
+
+	case *parser.NullCheck:
+		return evalNullCheck(t, env)
+
+	case *parser.ExistsExpr:
+		// The subquery within the exists should have been executed before
+		// expression evaluation and the exists nodes replaced with the result.
+
+	case parser.BytesVal:
+		v := string(t)
+		return sqlwire.Datum{StringVal: &v}, nil
+
+	case parser.StrVal:
+		v := string(t)
+		return sqlwire.Datum{StringVal: &v}, nil
+
+	case parser.IntVal:
+		v, err := strconv.ParseUint(string(t), 0, 64)
+		if err != nil {
+			return null, err
+		}
+		return sqlwire.Datum{UintVal: &v}, nil
+
+	case parser.NumVal:
+		v, err := strconv.ParseFloat(string(t), 64)
+		if err != nil {
+			return null, err
+		}
+		return sqlwire.Datum{FloatVal: &v}, nil
+
+	case parser.BoolVal:
+		v := bool(t)
+		return sqlwire.Datum{BoolVal: &v}, nil
+
+	case parser.ValArg:
+		// Placeholders should have been replaced before expression evaluation.
+
+	case *parser.NullVal:
+		return null, nil
+
+	case *parser.ColName:
+		if d, ok := env.Get(t.String()); ok {
+			return d, nil
+		}
+		return null, fmt.Errorf("column \"%s\" not found", t)
+
+	case parser.ValTuple:
+		if len(t) != 1 {
+			return null, fmt.Errorf("unsupported expression type: %T: %s", expr, expr)
+		}
+		return EvalExpr(t[0], env)
+
+	case *parser.Subquery:
+		// The subquery should have been executed before expression evaluation and
+		// the result placed into the expression tree.
+
+	case *parser.BinaryExpr:
+		return evalBinaryExpr(t, env)
+
+	case *parser.UnaryExpr:
+		return evalUnaryExpr(t, env)
+
+	case *parser.FuncExpr:
+		return evalFuncExpr(t, env)
+
+	case *parser.CaseExpr:
+		return evalCaseExpr(t, env)
+	}
+
+	return null, fmt.Errorf("unsupported expression type: %T", expr)
+}
+
+func evalAndExpr(expr *parser.AndExpr, env Env) (sqlwire.Datum, error) {
+	left, err := EvalExpr(expr.Left, env)
+	if err != nil {
+		return null, err
+	}
+	r := sqlwire.Datum{BoolVal: new(bool)}
+	if !left.Bool() {
+		return r, nil
+	}
+	right, err := EvalExpr(expr.Right, env)
+	if err != nil {
+		return null, err
+	}
+	if right.Bool() {
+		*r.BoolVal = true
+	}
+	return r, nil
+}
+
+func evalOrExpr(expr *parser.OrExpr, env Env) (sqlwire.Datum, error) {
+	left, err := EvalExpr(expr.Left, env)
+	if err != nil {
+		return null, err
+	}
+	r := sqlwire.Datum{BoolVal: new(bool)}
+	if left.Bool() {
+		*r.BoolVal = true
```

Comment:
This could use a helper function instead of mutating the local `r` variable:

```
func boolToDatum(b bool) sqlwire.Datum {
    return sqlwire.Datum{BoolVal: &b}
}
```